File: frontend/services/jasaAngkut.ts
import api from '../utils/api';
import { PaymentMethod } from './keuangan';

// --- Interfaces ---

export interface Supir {
    id: number;
    kode: string;
    nama: string;
    nik?: string;
    alamat?: string;
    telepon?: string;
    nomor_sim?: string;
    jenis_sim?: string;
    armada_default_id?: number;
    armada_default?: Armada;
    nopol_kendaraan?: string;
    info_kendaraan?: string;
    tanggal_bergabung: string;
    is_active: boolean;
    is_ready?: boolean;
    catatan?: string;
    created_at: string;
    updated_at: string;
}

export interface SupirSummary {
    supir: Supir;
    total_muatan: number;
    total_pendapatan: number;
    total_laba_tpm: number;
    total_piutang: number;
}

export interface Armada {
    id: number;
    nama: string;
    nopol: string;
    jenis?: string;
    is_active: boolean;
    is_ready?: boolean;
    catatan?: string;
    created_at: string;
    updated_at: string;
}

export interface Muatan {
    id: number;
    nomor_transaksi: string;
    tanggal: string;
    supir_id: number;
    supir?: Supir;
    supir_nama?: string;
    supir_nama_manual?: string;
    asal: string;
    tujuan: string;
    jenis_muatan?: string;
    berat_muatan?: string;
    armada_id?: number;
    armada?: Armada;
    nopol?: string;
    info_kendaraan?: string;
    ritase?: number;
    harga_beli?: number;
    harga_jual?: number;
    pendapatan_kotor: number;
    biaya_bbm: number;
    biaya_tol: number;
    biaya_makan: number;
    biaya_parkir: number;
    biaya_lainnya: number;
    total_biaya: number;
    laba_kotor: number;
    persentase_tpm: number;
    laba_tpm: number;
    laba_supir: number;
    status: 'PROSES' | 'SELESAI';
    status_bayar: 'LUNAS' | 'BELUM_LUNAS';
    metode_bayar?: PaymentMethod;
    tanggal_bayar?: string;
    biaya_tambahan?: BiayaLainnya[];
    part_services?: PartService[];
    piutang_id?: number;
    jumlah_bayar?: number;
    catatan?: string;
    created_at: string;
}

export interface MuatanSummary {
    total_transaksi?: number;
    lunas_count?: number;
    partial_count?: number;
    unpaid_count?: number;
    batal_count?: number;
    total_pendapatan?: number;
    laba_tpm?: number;
    saldo_bop?: number;
    total_tunai?: number;
    total_transfer?: number;
    total_dana_dari_utama?: number;
    details?: {
        gross_share_tpm: number;
        biaya_lainnya: number;
        biaya_bengkel: number;
    };
}

export interface MuatanCreate {
    tanggal: string; // YYYY-MM-DD
    supir_id: number;
    asal: string;
    tujuan: string;
    jenis_muatan?: string;
    berat_muatan?: string;
    pendapatan_kotor: number;
    biaya_bbm?: number;
    biaya_tol?: number;
    biaya_makan?: number;
    biaya_parkir?: number;
    biaya_lainnya?: number;
    persentase_tpm?: number; // Default 50
    status?: 'PROSES' | 'SELESAI';
    status_bayar?: 'LUNAS' | 'BELUM_LUNAS';
    metode_bayar?: PaymentMethod;
    biaya_operasional?: any[];
    catatan?: string;
}

export interface BiayaLainnya {
    id: number;
    muatan_id: number;
    kategori: string;
    deskripsi: string;
    jumlah: number;
    catatan?: string;
}

export interface PartService {
    id: number;
    tanggal: string;
    tipe: 'part' | 'service';
    deskripsi: string;
    qty: number;
    harga_satuan: number;
    total: number;
    catatan?: string;
}

// --- Service ---

export const jasaAngkutService = {
    // Supir Methods
    getSupirList: async (params?: {
        skip?: number;
        limit?: number;
        search?: string;
        is_active?: boolean;
        sort_by?: string;
        sort_order?: 'asc' | 'desc';
    }) => {
        const response = await api.get('/supir', { params });
        return response.data;
    },

    getActiveSupir: async () => {
        const response = await api.get('/supir/active');
        return response.data;
    },

    getSupir: async (id: number) => {
        const response = await api.get(`/supir/${id}`);
        return response.data;
    },

    createSupir: async (data: any) => {
        const response = await api.post('/supir', data);
        return response.data;
    },

    updateSupir: async (id: number, data: any) => {
        const response = await api.put(`/supir/${id}`, data);
        return response.data;
    },

    setSupirActive: async (id: number, isActive: boolean) => {
        const response = await api.patch(`/supir/${id}/active?is_active=${isActive}`);
        return response.data;
    },

    getSupirSummary: async (id: number) => {
        const response = await api.get(`/supir/${id}/summary`);
        return response.data;
    },

    deleteSupir: async (id: number) => {
        const response = await api.delete(`/supir/${id}`);
        return response.data;
    },

    // Armada Methods
    getArmadaList: async (params?: {
        skip?: number;
        limit?: number;
        search?: string;
        is_active?: boolean;
    }) => {
        const response = await api.get('/armada', { params });
        return response.data;
    },

    getActiveArmada: async (tanggal?: string) => {
        const response = await api.get('/armada/active', { params: { tanggal } });
        return response.data;
    },

    getArmada: async (id: number) => {
        const response = await api.get(`/armada/${id}`);
        return response.data;
    },

    createArmada: async (data: any) => {
        const response = await api.post('/armada', data);
        return response.data;
    },

    updateArmada: async (id: number, data: any) => {
        const response = await api.put(`/armada/${id}`, data);
        return response.data;
    },

    deleteArmada: async (id: number) => {
        const response = await api.delete(`/armada/${id}`);
        return response.data;
    },

    getArmadaDetail: async (id: number) => {
        const response = await api.get(`/armada/${id}/detail`);
        return response.data;
    },

    addArmadaExpense: async (id: number, data: {
        tanggal: string;
        deskripsi: string;
        jumlah: number;
        kategori?: string;
        catatan?: string;
        metode_bayar?: string;
        payments?: { metode: string; nominal: number; catatan?: string }[];
    }) => {
        const response = await api.post(`/armada/${id}/expense`, data);
        return response.data;
    },

    // Muatan Methods
    getMuatanList: async (params?: {
        skip?: number;
        limit?: number;
        search?: string;
        supir_id?: number;
        status_bayar?: string;
        tanggal_dari?: string;
        tanggal_sampai?: string;
    }) => {
        const response = await api.get('/muatan', { params });
        return response.data;
    },

    getMuatan: async (id: number) => {
        const response = await api.get(`/muatan/${id}`);
        return response.data;
    },

    createMuatan: async (data: MuatanCreate) => {
        const response = await api.post('/muatan', data);
        return response.data;
    },

    updateMuatan: async (id: number, data: any) => {
        const response = await api.put(`/muatan/${id}`, data);
        return response.data;
    },

    updateMuatanStatus: async (id: number, status: 'PROSES' | 'SELESAI') => {
        const response = await api.patch(`/muatan/${id}/status`, null, { params: { status } });
        return response.data;
    },

    deleteMuatan: async (id: number) => {
        const response = await api.delete(`/muatan/${id}`);
        return response.data;
    },

    getMuatanSummary: async (params?: { tanggal_dari?: string; tanggal_sampai?: string; search?: string }): Promise<MuatanSummary> => {
        const response = await api.get('/muatan/summary', { params });
        return response.data;
    },

    getRouteSuggestions: async (field: 'asal' | 'tujuan', query?: string, limit: number = 10): Promise<string[]> => {
        const response = await api.get(`/muatan/suggestions/${field}`, { params: { q: query, limit } });
        return response.data;
    },


    getDriverRecentTrips: async (supirId: number, limit: number = 5) => {
        const response = await api.get(`/muatan/supir/${supirId}/recent`, { params: { limit } });
        return response.data;
    },

    getMuatanBySupir: async (supirId: number) => {
        const response = await api.get(`/muatan/supir/${supirId}`);
        return response.data;
    },

    markPaid: async (id: number, metode_bayar: string = 'tunai') => {
        const response = await api.patch(`/muatan/${id}/paid`, null, {
            params: { metode_bayar: metode_bayar.toUpperCase() }
        });
        return response.data;
    },

    payMuatanSplit: async (data: { muatan_id: number; tanggal: string; payments: { metode: string; nominal: number; catatan?: string }[]; catatan?: string }) => {
        const response = await api.post(`/muatan/${data.muatan_id}/paid-split`, data);
        return response.data;
    },

    addMuatanCost: async (muatanId: number, data: any) => {
        const response = await api.post(`/muatan/${muatanId}/costs`, data);
        return response.data;
    },
};

File: backend/app/schemas/jasa_angkut.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List

from pydantic import BaseModel, Field

from app.utils.constants import PaymentStatus, PaymentMethod, MuatanStatus, KasBankJenis
from app.schemas.keuangan import PaymentDetail, PembayaranPiutangResponse
from app.schemas.bengkel import PengeluaranBengkelResponse


# ============================================
# ARMADA (VEHICLE) SCHEMAS
# ============================================

class ArmadaBase(BaseModel):
    """Base armada schema."""

    nama: str = Field(..., min_length=2, max_length=100)
    nopol: str = Field(..., min_length=2, max_length=20)
    jenis: Optional[str] = Field(None, max_length=50)
    is_active: bool = True
    catatan: Optional[str] = None


class ArmadaCreate(ArmadaBase):
    """Schema for creating armada."""
    pass


class ArmadaExpenseCreate(BaseModel):
    """Schema for adding an expense to an armada."""
    tanggal: date
    kategori: str = "Operasional"
    deskripsi: str = Field(..., min_length=1, max_length=255)
    jumlah: Decimal = Field(..., ge=0)
    catatan: Optional[str] = None
    metode_bayar: Optional[PaymentMethod] = PaymentMethod.TUNAI
    kas_jenis: Optional[KasBankJenis] = None
    payments: Optional[List[PaymentDetail]] = None



class ArmadaUpdate(BaseModel):
    """Schema for updating armada."""

    nama: Optional[str] = Field(None, min_length=2, max_length=100)
    nopol: Optional[str] = Field(None, min_length=2, max_length=20)
    jenis: Optional[str] = Field(None, max_length=50)
    is_active: Optional[bool] = None
    catatan: Optional[str] = None


class ArmadaResponse(ArmadaBase):
    """Schema for armada response."""

    id: int
    is_ready: Optional[bool] = True  # Calculated field
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ArmadaList(BaseModel):
    """Schema for paginated armada list."""

    data: List[ArmadaResponse]
    total: int
    page: int
    size: int
    pages: int


# ============================================
# SUPIR (DRIVER) SCHEMAS
# ============================================

class SupirBase(BaseModel):
    """Base driver schema."""

    nama: str = Field(..., min_length=2, max_length=100)
    nik: Optional[str] = Field(None, max_length=20)
    alamat: Optional[str] = None
    telepon: Optional[str] = Field(None, max_length=20)
    nomor_sim: Optional[str] = Field(None, max_length=30)
    jenis_sim: Optional[str] = Field(None, max_length=10)
    nopol_kendaraan: Optional[str] = Field(None, max_length=20)
    info_kendaraan: Optional[str] = Field(None, max_length=255)
    catatan: Optional[str] = None


class SupirCreate(SupirBase):
    """Schema for creating driver."""

    kode: Optional[str] = Field(None, max_length=20)
    tanggal_bergabung: date
    armada_default_id: Optional[int] = None


class SupirUpdate(BaseModel):
    """Schema for updating driver."""

    nama: Optional[str] = Field(None, min_length=2, max_length=100)
    nik: Optional[str] = Field(None, max_length=20)
    alamat: Optional[str] = None
    telepon: Optional[str] = Field(None, max_length=20)
    nomor_sim: Optional[str] = Field(None, max_length=30)
    jenis_sim: Optional[str] = Field(None, max_length=10)
    nopol_kendaraan: Optional[str] = Field(None, max_length=20)
    info_kendaraan: Optional[str] = Field(None, max_length=255)
    is_active: Optional[bool] = None
    catatan: Optional[str] = None


class SupirResponse(BaseModel):
    """Schema for driver response."""

    id: int
    kode: str
    nama: str
    nik: Optional[str] = None
    alamat: Optional[str] = None
    telepon: Optional[str] = None
    nomor_sim: Optional[str] = None
    jenis_sim: Optional[str] = None
    nopol_kendaraan: Optional[str] = None
    info_kendaraan: Optional[str] = None
    armada_default_id: Optional[int] = None
    armada_default: Optional[ArmadaResponse] = None
    tanggal_bergabung: date
    is_active: bool
    catatan: Optional[str] = None
    is_ready: Optional[bool] = True
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SupirList(BaseModel):
    """Schema for paginated driver list."""

    data: List[SupirResponse]
    total: int
    page: int
    size: int
    pages: int

    total_piutang: Decimal = Decimal("0")


# ============================================
# MUATAN JASA ANGKUT SCHEMAS
# ============================================

class BiayaItem(BaseModel):
    """Schema for dynamic operational cost item."""
    id: Optional[int] = None # For updates
    deskripsi: str = Field(..., min_length=1, max_length=255)
    jumlah: Decimal = Field(..., ge=0)


class BiayaTambahanResponse(BaseModel):
    """Schema for additional cost response."""
    id: int
    kategori: str
    deskripsi: str
    jumlah: Decimal
    model_config = {"from_attributes": True}


class PartServiceResponse(BaseModel):
    """Schema for part/service cost response (from linked bengkel transactions)."""
    id: int
    tanggal: date
    tipe: str  # 'part' or 'service'
    deskripsi: str
    qty: int
    harga_satuan: Decimal
    total: Decimal
    catatan: Optional[str] = None
    model_config = {"from_attributes": True}

class MuatanCreate(BaseModel):
    """Schema for creating transport load."""

    tanggal: date
    supir_id: Optional[int] = None
    supir_nama: Optional[str] = Field(None, max_length=1000)
    armada_id: Optional[int] = None
    nopol: Optional[str] = Field(None, max_length=20)
    info_kendaraan: Optional[str] = Field(None, max_length=1000)
    asal: str = Field(..., min_length=2, max_length=1000)
    tujuan: str = Field(..., min_length=2, max_length=1000)
    jenis_muatan: Optional[str] = Field(None, max_length=1000)
    ritase: int = Field(default=1, ge=1)
    berat_muatan: Optional[str] = Field(None, max_length=1000)
    
    # Trading values
    harga_beli: Decimal = Field(default=Decimal("0"), ge=0)
    harga_jual: Decimal = Field(default=Decimal("0"), ge=0)
    
    # Calculated on backend
    pendapatan_kotor: Optional[Decimal] = Field(None, ge=0)

    # Dynamic operational costs
    biaya_operasional: List[BiayaItem] = []
    
    # Legacy fields (kept for backward compatibility but default to 0)
    biaya_bbm: Decimal = Field(default=Decimal("0"), ge=0)
    biaya_tol: Decimal = Field(default=Decimal("0"), ge=0)
    biaya_makan: Decimal = Field(default=Decimal("0"), ge=0)
    biaya_parkir: Decimal = Field(default=Decimal("0"), ge=0)
    biaya_lainnya: Decimal = Field(default=Decimal("0"), ge=0)
    
    persentase_tpm: Decimal = Field(default=Decimal("100"), ge=0, le=100)
    status: Optional[MuatanStatus] = MuatanStatus.PROSES
    status_bayar: Optional[PaymentStatus] = PaymentStatus.BELUM_LUNAS
    metode_bayar: Optional[PaymentMethod] = PaymentMethod.TUNAI
    kas_jenis: Optional[KasBankJenis] = None
    payments: Optional[List[PaymentDetail]] = None
    catatan: Optional[str] = None


class MuatanUpdate(BaseModel):
    """Schema for updating transport load."""

    tanggal: Optional[date] = None
    supir_id: Optional[int] = None
    supir_nama: Optional[str] = Field(None, max_length=100)
    armada_id: Optional[int] = None
    nopol: Optional[str] = Field(None, max_length=20)
    info_kendaraan: Optional[str] = Field(None, max_length=255)
    asal: Optional[str] = Field(None, min_length=2, max_length=1000)
    tujuan: Optional[str] = Field(None, min_length=2, max_length=1000)
    jenis_muatan: Optional[str] = Field(None, max_length=1000)
    ritase: Optional[int] = Field(None, ge=1)
    berat_muatan: Optional[str] = Field(None, max_length=50)
    
    harga_beli: Optional[Decimal] = Field(None, ge=0)
    harga_jual: Optional[Decimal] = Field(None, ge=0)
    pendapatan_kotor: Optional[Decimal] = Field(None, ge=0)

    # Dynamic operational costs
    biaya_operasional: Optional[List[BiayaItem]] = None
    
    # Legacy fields
    biaya_bbm: Optional[Decimal] = Field(None, ge=0)
    biaya_tol: Optional[Decimal] = Field(None, ge=0)
    biaya_makan: Optional[Decimal] = Field(None, ge=0)
    biaya_parkir: Optional[Decimal] = Field(None, ge=0)
    biaya_lainnya: Optional[Decimal] = Field(None, ge=0)
    
    persentase_tpm: Optional[Decimal] = None
    status: Optional[MuatanStatus] = None
    status_bayar: Optional[PaymentStatus] = None
    metode_bayar: Optional[PaymentMethod] = None
    payments: Optional[List[PaymentDetail]] = None
    kas_jenis: Optional[KasBankJenis] = None
    catatan: Optional[str] = None


class MuatanResponse(BaseModel):
    """Schema for transport load response."""

    id: int
    nomor_transaksi: str
    tanggal: date
    supir_id: Optional[int] = None
    supir: Optional[SupirResponse] = None # Full driver object
    supir_nama: Optional[str] = None
    supir_nama_manual: Optional[str] = None
    armada_id: Optional[int] = None
    armada: Optional[ArmadaResponse] = None
    nopol: Optional[str] = None
    info_kendaraan: Optional[str] = None
    asal: str
    tujuan: str
    jenis_muatan: Optional[str] = None
    ritase: int
    berat_muatan: Optional[str] = None
    
    harga_beli: Decimal
    harga_jual: Decimal
    pendapatan_kotor: Decimal
    
    biaya_bbm: Decimal
    biaya_tol: Decimal
    biaya_makan: Decimal
    biaya_parkir: Decimal
    biaya_lainnya: Decimal
    total_biaya: Decimal
    laba_kotor: Decimal
    persentase_tpm: Decimal
    laba_tpm: Decimal
    laba_supir: Decimal
    status: MuatanStatus
    status_bayar: PaymentStatus
    tanggal_bayar: Optional[date] = None
    biaya_tambahan: List[BiayaTambahanResponse] = []
    part_services: List[PartServiceResponse] = []
    piutang_id: Optional[int] = None
    jumlah_bayar: Decimal = Decimal("0")
    payment_history: List[PembayaranPiutangResponse] = []
    catatan: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MuatanPaymentSplit(BaseModel):
    """Schema for split payment of transport load."""
    muatan_id: int
    tanggal: date
    payments: List[PaymentDetail]
    catatan: Optional[str] = None


class MuatanList(BaseModel):
    """Schema for paginated transport load list."""

    data: List[MuatanResponse]
    total: int
    page: int
    size: int
    pages: int
from app.schemas.bengkel import TransaksiBengkelResponse


class ArmadaStats(BaseModel):
    """Aggregated stats for armada."""
    total_muatan: int = 0
    total_ritase: int = 0
    total_pendapatan_kotor: Decimal = Decimal("0")
    total_biaya_operasional: Decimal = Decimal("0")
    total_perbaikan_bengkel: Decimal = Decimal("0")
    total_laba_tpm: Decimal = Decimal("0")


class ArmadaDetailResponse(BaseModel):
    """Detailed response for armada including history and stats."""
    armada: ArmadaResponse
    stats: ArmadaStats
    muatan_history: List[MuatanResponse]
    perbaikan_history: List[TransaksiBengkelResponse]
    general_expenses: List[BiayaTambahanResponse] = []
    workshop_expenses: List[PengeluaranBengkelResponse] = []


    model_config = {"from_attributes": True}
